### PR TITLE
copr: Add vectorize sha2

### PR DESCRIPTION
**Pull request by Rustin-Liu**
Signed-off-by: Rustin-Liu <rustin.liu@gmail.com>
PCP #5751

###  What have you changed?
Add vectorize sha2.

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?
- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

None

###  Does this PR affect `tidb-ansible`?

None

###  Refer to a related PR or issue link (optional)
#5751

###  Benchmark result if necessary (optional)
None
###  Any examples? (optional)
None

### CODE REVIEW

**Comment by breeswish** [comment]
Thanks a lot! Please also run your PR against a copr-test that enables the sha2 push down.

**Comment by Rustin-Liu** [comment]
> Thanks a lot! Please also run your PR against a copr-test that enables the sha2 push down.

passed.

**Comment by Rustin-Liu** [comment]
@sre-bot  /run-integration-copr-test copr-test=pr/98

File: components/tidb_query/src/rpn_expr/impl_encryption.rs [hunk 1]
```diff
@@ -117,4 +144,41 @@ mod tests {
         }
         test_unary_func_ok_none::<Bytes, Bytes>(ScalarFuncSig::Sha1);
     }
+
+    #[test]
+    fn test_sha2() {
+        let cases = vec![
```
Comment (rleungx) [inline components/tidb_query/src/rpn_expr/impl_encryption.rs]: Can we add some NULL cases?

Comment (Rustin-Liu) [reply components/tidb_query/src/rpn_expr/impl_encryption.rs]: added.

**Comment by Rustin-Liu** [comment]
@sre-bot   /run-integration-copr-test copr-test=pr/98

**Comment by Rustin-Liu** [comment]
@rleungx  PTAL~

File: components/tidb_query/src/rpn_expr/impl_encryption.rs [hunk 2]
```diff
@@ -26,6 +32,26 @@ pub fn sha1(arg: &Option<Bytes>) -> Result<Option<Bytes>> {
...
+pub fn sha2(input: &Option<Bytes>, hash_length: &Option<Int>) -> Result<Option<Bytes>> {
+    match (input, hash_length) {
+        (Some(input), Some(hash_length)) => {
+            let sha2 = match *hash_length {
+                SHA0 | SHA256 => MessageDigest::sha256(),
+                SHA224 => MessageDigest::sha224(),
+                SHA384 => MessageDigest::sha384(),
+                SHA512 => MessageDigest::sha512(),
+                _ => {
+                    return Err(Error::incorrect_parameters("sha2").into());
```
Comment (niedhui) [inline components/tidb_query/src/rpn_expr/impl_encryption.rs]: There should be a warning in context and return `Ok(None)` when  `hash_length`  is invalid.

Comment (Rustin-Liu) [reply components/tidb_query/src/rpn_expr/impl_encryption.rs]: Could you please explain why we can not return err?
I think the parameters incorrect should return err.

Comment (niedhui) [reply components/tidb_query/src/rpn_expr/impl_encryption.rs]: We need the same behavior as MySQL does.

https://dev.mysql.com/doc/refman/5.5/en/encryption-functions.html#function_sha2

>  If either argument is NULL or the hash length is not one of the permitted values, the return value is NULL

and you can check the behavior in MySQL console
```
mysql> select sha2("abc", 12);
+-----------------+
| sha2("abc", 12) |
+-----------------+
| NULL            |
+-----------------+
1 row in set, 1 warning (0.01 sec)

mysql> show warnings;
+---------+------+------------------------------------------------------------+
| Level   | Code | Message                                                    |
+---------+------+------------------------------------------------------------+
| Warning | 1583 | Incorrect parameters in the call to native function 'sha2' |
+---------+------+------------------------------------------------------------+
1 row in set (0.00 sec)
```

Comment (Rustin-Liu) [reply components/tidb_query/src/rpn_expr/impl_encryption.rs]: Thank  you, got it. 
I will add it tonight.

Comment (Rustin-Liu) [reply components/tidb_query/src/rpn_expr/impl_encryption.rs]: modified

**Comment by Rustin-Liu** [comment]
@sre-bot   /run-integration-copr-test copr-test=pr/98

**Comment by Rustin-Liu** [comment]
Passed all tests. 
PTAL~

**Comment by niedhui** [comment]
LGTM. Thanks

**Comment by Rustin-Liu** [comment]
@sre-bot  /run-auto-merge//merge

**Comment by Rustin-Liu** [comment]
@sre-bot  /merge

**Comment by breeswish** [comment]
Sorry, the integration test is broken recently. Your PR will be merged after it is fixed :)

**Comment by Rustin-Liu** [comment]
@sre-bot /merge

**Pull request merged**